File: frontend/src/api/ile.ts
/**
 * Interactive Learning Engine (Flash Tutor) API client — CB-ILE-001
 */
import { api } from './client';

// --- Types ---

export interface ILEQuestionOption {
  A: string;
  B: string;
  C: string;
  D: string;
}

export interface ILEQuestion {
  index: number;
  question: string;
  options: ILEQuestionOption | null;
  format: string;
  difficulty: string;
  blooms_tier: string;
}

export interface ILECurrentQuestion {
  session_id: number;
  question: ILEQuestion;
  question_index: number;
  total_questions: number;
  mode: string;
  attempt_number: number;
  disabled_options: string[];
  streak_count: number;
}

export interface ILESession {
  id: number;
  student_id: number;
  parent_id: number | null;
  mode: string;
  subject: string;
  topic: string;
  grade_level: number | null;
  question_count: number;
  difficulty: string;
  blooms_tier: string;
  timer_enabled: boolean;
  timer_seconds: number | null;
  is_private_practice: boolean;
  status: string;
  current_question_index: number;
  score: number | null;
  total_correct: number | null;
  xp_awarded: number | null;
  started_at: string;
  completed_at: string | null;
  expires_at: string | null;
  course_id: number | null;
  course_content_id: number | null;
}

export interface ILESessionSummary {
  id: number;
  mode: string;
  subject: string;
  topic: string;
  status: string;
  score: number | null;
  question_count: number;
  total_correct: number | null;
  xp_awarded: number | null;
  started_at: string;
  completed_at: string | null;
}

export interface ILEAnswerFeedback {
  is_correct: boolean;
  attempt_number: number;
  xp_earned: number;
  hint: string | null;
  parent_hint_note: string | null;
  explanation: string | null;
  correct_answer: string | null;
  question_complete: boolean;
  session_complete: boolean;
  streak_count: number;
  streak_broken: boolean;
}

export interface ILEQuestionResult {
  index: number;
  question: string;
  correct_answer: string;
  student_answer: string | null;
  is_correct: boolean;
  attempts: number;
  xp_earned: number;
  difficulty: string;
  format: string;
}

export interface ILEAreaToRevisit {
  index: number;
  question: string;
  correct_answer: string;
  student_answer: string | null;
  attempts: number;
}

export interface ILESessionResults {
  session_id: number;
  mode: string;
  subject: string;
  topic: string;
  score: number;
  total_questions: number;
  percentage: number;
  total_xp: number;
  questions: ILEQuestionResult[];
  streak_at_end: number;
  time_taken_seconds: number | null;
  weak_areas: string[];
  suggested_next_topic: string | null;
  areas_to_revisit: ILEAreaToRevisit[];
  aha_detected: boolean;
}

export interface ILETopic {
  subject: string;
  topic: string;
  course_id: number | null;
  course_name: string | null;
  mastery_pct: number | null;
  is_weak_area: boolean;
  next_review_at: string | null;
}

export interface ILEMasteryEntry {
  subject: string;
  topic: string;
  total_sessions: number;
  avg_attempts: number;
  is_weak_area: boolean;
  current_difficulty: string;
  last_score_pct: number | null;
  next_review_at: string | null;
  glow_intensity: number;
}

export interface ILEMasteryMap {
  student_id: number;
  entries: ILEMasteryEntry[];
  total_topics: number;
  mastered_topics: number;
  weak_topics: number;
}

export interface ILECareerConnect {
  career: string;
  connection: string;
}

export interface ILESessionCreate {
  mode: 'learning' | 'testing' | 'parent_teaching';
  subject: string;
  topic: string;
  grade_level?: number;
  question_count?: number;
  difficulty?: 'easy' | 'medium' | 'challenging';
  blooms_tier?: 'recall' | 'understand' | 'apply';
  timer_enabled?: boolean;
  timer_seconds?: number;
  is_private_practice?: boolean;
  course_id?: number;
  course_content_id?: number;
  child_student_id?: number;
}

// --- API ---

export const ileApi = {
  // Sessions
  createSession: (data: ILESessionCreate) =>
    api.post<ILESession>('/api/ile/sessions', data).then(r => r.data),

  createSessionFromStudyGuide: (params: {
    study_guide_id?: number;
    course_content_id?: number;
    mode?: string;
    question_count?: number;
    difficulty?: string;
  }) =>
    api.post<ILESession>('/api/ile/sessions/from-study-guide', null, { params }).then(r => r.data),

  getActiveSession: () =>
    api.get<ILESession | null>('/api/ile/sessions/active').then(r => r.data),

  getSession: (id: number) =>
    api.get<ILESession>(`/api/ile/sessions/${id}`).then(r => r.data),

  getSessionHistory: (limit = 20) =>
    api.get<ILESessionSummary[]>('/api/ile/sessions', { params: { limit } }).then(r => r.data),

  // Questions
  getCurrentQuestion: (sessionId: number) =>
    api.get<ILECurrentQuestion>(`/api/ile/sessions/${sessionId}/question`).then(r => r.data),

  submitAnswer: (sessionId: number, answer: string, timeTakenMs?: number, parentHintNote?: string) =>
    api.post<ILEAnswerFeedback>(`/api/ile/sessions/${sessionId}/answer`, {
      answer,
      time_taken_ms: timeTakenMs,
      parent_hint_note: parentHintNote || undefined,
    }).then(r => r.data),

  addParentHint: (sessionId: number, hintNote: string) =>
    api.post<{ question_index: number; parent_hint_note: string }>(`/api/ile/sessions/${sessionId}/parent-hint`, {
      hint_note: hintNote,
    }).then(r => r.data),

  // Session lifecycle
  completeSession: (sessionId: number) =>
    api.post<ILESessionResults>(`/api/ile/sessions/${sessionId}/complete`).then(r => r.data),

  abandonSession: (sessionId: number) =>
    api.post(`/api/ile/sessions/${sessionId}/abandon`).then(r => r.data),

  getSessionResults: (sessionId: number) =>
    api.get<ILESessionResults>(`/api/ile/sessions/${sessionId}/results`).then(r => r.data),

  // Topics
  getTopics: () =>
    api.get<{ topics: ILETopic[] }>('/api/ile/topics').then(r => r.data.topics),

  // Mastery
  getMasteryMap: () =>
    api.get<ILEMasteryMap>('/api/ile/mastery').then(r => r.data),

  getSurpriseMe: () =>
    api.get<{ topic: ILETopic; reason: string }>('/api/ile/topics/surprise-me').then(r => r.data),

  // Career Connect
  getCareerConnect: (sessionId: number) =>
    api.get<ILECareerConnect>(`/api/ile/sessions/${sessionId}/career-connect`).then(r => r.data),
};

File: app/schemas/ile.py
"""Pydantic schemas for the Interactive Learning Engine (CB-ILE-001)."""
from datetime import datetime

from pydantic import BaseModel, Field


# --- Session ---

class ILESessionCreate(BaseModel):
    mode: str = Field(..., pattern="^(learning|testing|parent_teaching)$")
    subject: str = Field(..., min_length=1, max_length=100)
    topic: str = Field(..., min_length=1, max_length=200)
    grade_level: int | None = None
    question_count: int = Field(default=5, ge=3, le=7)
    difficulty: str = Field(default="medium", pattern="^(easy|medium|challenging)$")
    blooms_tier: str = Field(default="recall", pattern="^(recall|understand|apply)$")
    timer_enabled: bool = False
    timer_seconds: int | None = Field(default=None, ge=30, le=3600)
    is_private_practice: bool = False
    course_id: int | None = None
    course_content_id: int | None = None
    # Parent teaching mode
    child_student_id: int | None = None


class ILESessionResponse(BaseModel):
    id: int
    student_id: int
    parent_id: int | None
    mode: str
    subject: str
    topic: str
    grade_level: int | None
    question_count: int
    difficulty: str
    blooms_tier: str
    timer_enabled: bool
    timer_seconds: int | None
    is_private_practice: bool
    status: str
    current_question_index: int
    score: int | None
    total_correct: int | None
    xp_awarded: int | None
    started_at: datetime
    completed_at: datetime | None
    expires_at: datetime | None
    course_id: int | None
    course_content_id: int | None

    class Config:
        from_attributes = True


class ILESessionSummary(BaseModel):
    id: int
    mode: str
    subject: str
    topic: str
    status: str
    score: int | None
    question_count: int
    total_correct: int | None
    xp_awarded: int | None
    started_at: datetime
    completed_at: datetime | None

    class Config:
        from_attributes = True


# --- Question ---

class ILEQuestionOption(BaseModel):
    A: str
    B: str
    C: str
    D: str


class ILEQuestion(BaseModel):
    index: int
    question: str
    options: ILEQuestionOption | None = None  # None for fill_blank
    format: str = "mcq"
    difficulty: str = "medium"
    blooms_tier: str = "recall"


class ILECurrentQuestion(BaseModel):
    session_id: int
    question: ILEQuestion
    question_index: int
    total_questions: int
    mode: str
    attempt_number: int = 1
    disabled_options: list[str] = []  # Previously wrong answers (Learning Mode)
    streak_count: int = 0


# --- Answer ---

class ILEAnswerSubmit(BaseModel):
    answer: str = Field(..., min_length=1, max_length=500)
    time_taken_ms: int | None = Field(default=None, ge=0)
    parent_hint_note: str | None = Field(default=None, max_length=500)


class ILEAnswerFeedback(BaseModel):
    is_correct: bool
    attempt_number: int
    xp_earned: int
    # Learning Mode fields
    hint: str | None = None
    parent_hint_note: str | None = None  # Parent Teaching Mode: parent's personal hint
    explanation: str | None = None
    correct_answer: str | None = None  # Revealed after max attempts or correct
    # Session progress
    question_complete: bool  # True if correct or auto-revealed
    session_complete: bool
    streak_count: int = 0
    streak_broken: bool = False
    difficulty_changed: str | None = None  # New difficulty level if adjusted


# --- Results ---

class ILEQuestionResult(BaseModel):
    index: int
    question: str
    correct_answer: str
    student_answer: str | None
    is_correct: bool
    attempts: int
    xp_earned: int
    difficulty: str
    format: str


class ILEAreaToRevisit(BaseModel):
    index: int
    question: str
    correct_answer: str
    student_answer: str | None
    attempts: int


class ILESessionResults(BaseModel):
    session_id: int
    mode: str
    subject: str
    topic: str
    score: int
    total_questions: int
    percentage: float
    total_xp: int
    questions: list[ILEQuestionResult]
    streak_at_end: int
    time_taken_seconds: int | None
    # Adaptive feedback
    weak_areas: list[str] = []
    suggested_next_topic: str | None = None
    areas_to_revisit: list[ILEAreaToRevisit] = []
    aha_detected: bool = False


# --- Career Connect ---

class ILECareerConnect(BaseModel):
    career: str
    connection: str


# --- Topics ---

class ILETopic(BaseModel):
    subject: str
    topic: str
    course_id: int | None = None
    course_name: str | None = None
    mastery_pct: float | None = None
    is_weak_area: bool = False
    next_review_at: datetime | None = None


class ILETopicList(BaseModel):
    topics: list[ILETopic]


class ILESurpriseMe(BaseModel):
    topic: ILETopic
    reason: str  # e.g. "Weak area — avg 2.3 attempts"


# --- Mastery ---

class ILEMasteryEntry(BaseModel):
    subject: str
    topic: str
    total_sessions: int
    avg_attempts: float
    is_weak_area: bool
    current_difficulty: str
    last_score_pct: float | None
    next_review_at: datetime | None
    glow_intensity: float = 1.0  # 0.0 (faded) to 1.0 (bright)

    class Config:
        from_attributes = True


class ILEMasteryMap(BaseModel):
    student_id: int
    entries: list[ILEMasteryEntry]
    total_topics: int
    mastered_topics: int
    weak_topics: int


# --- Parent Teaching Mode ---

class ILEParentHintSubmit(BaseModel):
    hint_note: str = Field(..., min_length=1, max_length=500)


class ILEParentHintResponse(BaseModel):
    question_index: int
    parent_hint_note: str


# --- Admin Analytics (#3216) ---

class ILEDailySessionCount(BaseModel):
    date: str
    count: int


class ILEModeSplit(BaseModel):
    mode: str
    count: int


class ILETopTopic(BaseModel):
    topic: str
    count: int


class ILEAdminAnalytics(BaseModel):
    sessions_per_day: list[ILEDailySessionCount]
    total_sessions: int
    completed_sessions: int
    completion_rate: float
    average_score: float | None
    average_cost_per_session: float | None
    mode_split: list[ILEModeSplit]
    top_topics: list[ILETopTopic]
    flagged_sessions: int
